finality-provider/service/fp_instance.go

```diff
@@ -194,52 +193,51 @@ func (fp *FinalityProviderInstance) IsRunning() bool {
 func (fp *FinalityProviderInstance) finalitySigSubmissionLoop() {
 	defer fp.wg.Done()
 
+	var targetHeight uint64
 	for {
 		select {
 		case b := <-fp.poller.GetBlockInfoChan():
-			fp.logger.Debug(
-				"the finality-provider received a new block, start processing",
-				zap.String("pk", fp.GetBtcPkHex()),
-				zap.Uint64("height", b.Height),
-				zap.String("block_hash", hex.EncodeToString(b.Hash)),
-			)
-
-			// check whether the block has been processed before
-			if fp.hasProcessed(b.Height) {
-				continue
-			}
-			// check whether the finality provider has voting power
-			hasVp, err := fp.hasVotingPower(b.Height)
-			if err != nil {
-				fp.reportCriticalErr(err)
-				continue
-			}
-			if !hasVp {
-				// the finality provider does not have voting power
-				// and it will never will at this block
-				fp.MustSetLastProcessedHeight(b.Height)
-				fp.metrics.IncrementFpTotalBlocksWithoutVotingPower(fp.GetBtcPkHex())
-				continue
-			}
-			// check whether the randomness has been committed
-			// the retry will end if max retry times is reached
-			// or the target block is finalized
-			isFinalized, err := fp.retryCheckRandomnessUntilBlockFinalized(b)
+			shouldProcess, err := fp.shouldProcessBlock(b)
 			if err != nil {
 				if !errors.Is(err, ErrFinalityProviderShutDown) {
 					fp.reportCriticalErr(err)
 				}
-				break
+				continue
 			}
-			// the block is finalized, no need to submit finality signature
-			if isFinalized {
-				fp.MustSetLastProcessedHeight(b.Height)
+			if !shouldProcess {
 				continue
 			}
-
-			// use the copy of the block to avoid the impact to other receivers
-			nextBlock := *b
-			res, err := fp.retrySubmitFinalitySignatureUntilBlockFinalized(&nextBlock)
+			// Fetch all available blocks
+			// Note: not all the blocks in the range will have votes cast
+			// due to lack of voting power or public randomness, so we may
+			// have gaps during processing
+			pollerBlocks := []*types.BlockInfo{b}
+			fetchMoreBlocks := true
```

Comment:
we probably don't need this flag. We can use `for {}` and inject `break` in `default` case

Reply:
yes, it is removed.